### PR TITLE
Algorithm integration

### PR DESCRIPTION
**Pull request by ranzmigrod**

### CODE REVIEW

File: src/DroneSafetyApi/Program.cs [hunk 1]
```diff
@@ -22,4 +22,4 @@ public static void Main(string[] args)
             host.Run();
         }
     }
-}
+}
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Program.cs]: Why did you do this?

File: src/DroneSafetyApi/Services/HazardsToHeatmapResponse.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+namespace DroneSafetyApi.Services
+{
+    public class HazardsToHeatmapResponse : HazardsToHeatmapResponseStrategy
+    {
+        public override Heatmap ConvertToHeatmap(int height, int width, BoundingBox area, IEnumerable<Hazard> hazards)
+        {
+            // Initialise Heatmap
+            Heatmap heatmap = new Heatmap();
+            heatmap.Values = new int[height][];
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Services/HazardsToHeatmapResponse.cs]: Change the code to [height, width]. Also, fix this elsewhere if it's an issue.

File: src/DroneSafetyApi/Services/HazardsToHeatmapResponse.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            for(int i = 0; i < height; i++)
+            {
+                heatmap.Values[i] = new int[width];
+                for(int j = 0; j < width; j++)
+                {
+                    heatmap.Values[i][j] = 0;
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Services/HazardsToHeatmapResponse.cs]: It automatically initialises itself to zero

File: src/DroneSafetyApi/Services/HazardsToHeatmapResponse.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using DroneSafetyApi.Models;
+using Microsoft.Azure.Documents.Spatial;
+
+namespace DroneSafetyApi.Services
+{
+    public class HazardsToHeatmapResponse : HazardsToHeatmapResponseStrategy
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Services/HazardsToHeatmapResponse.cs]: This class doesn't exist anymore?

File: src/DroneSafetyApi/Models/DataPoint.cs [hunk 5]
```diff
@@ -7,11 +7,13 @@
...
 namespace DroneSafetyApi.Models
 {
-    public class Hazard
+    public class DataPoint
     {
         [JsonProperty(PropertyName = "id")]
-        public string Id { get; set; }
-        public int Severity { get; set; }
-        public Polygon Area { get; set; }
+        public string DataType { get; set; }
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Models/DataPoint.cs]: DataType should not have PropertyName "id" as it will not necessarily be unique across different instances in the collection. It's probably good practice to leave in the "Id" field, so don't delete the property; re-add the "Id" field.

File: src/DroneSafetyApi/Models/DataPoint.cs [hunk 6]
```diff
@@ -7,11 +7,13 @@
...
+        public int Time { get; set; }
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Models/DataPoint.cs]: Why are we using an "int" for time as opposed to something like DateTime?

File: src/DroneSafetyApi/Models/DataPoint.cs [hunk 7]
```diff
@@ -7,11 +7,13 @@
...
+        public Geometry Location { get; set; }
+        public int Data_ID { get; set; }
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Models/DataPoint.cs]: What is this field for?

File: src/DroneSafetyApi/Models/HeatMapPoint.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace DroneSafetyApi.Models
+{
+    public class HeatMapPoint
+    {
+        public double x { get; set; }
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Models/HeatMapPoint.cs:10]: Should Have Capital letters for properties of classes. Are x and y good names, or would  we prefer Latitude and Longitude?

File: src/DroneSafetyApi/Models/HeatMapPoint.cs [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+        public double y { get; set; }
+
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Models/HeatMapPoint.cs:12]: Why all these extra newlines?

File: src/DroneSafetyApi/Models/Heatmap.cs [hunk 10]
```diff
@@ -1,12 +1,81 @@
...
 using System.Threading.Tasks;
 
 namespace DroneSafetyApi.Models
 {
-    public class Heatmap
+
+    public class HeatMap
     {
-        public int[][] Values { get; set; }
+        private int[,] mHeatMap;
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Models/Heatmap.cs:13]: We're not using the 'prefix member variables with "m"' convention

File: src/DroneSafetyApi/Program.cs (hunk 1 above)
Comment (joshuablake) [inline src/DroneSafetyApi/Program.cs:25]: It's removing the newline at end of file.

File: src/DroneSafetyApi/Models/Heatmap.cs [hunk 11]
```diff
@@ -1,12 +1,81 @@
...
+        private double mStartX;
+        private double mStartY;
+        private int mDecimalPlaces;
+        private double mResolution;
+
+        public double GetResolution()
+        {
+            return mResolution;
+        }
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Models/Heatmap.cs:24]: Why do we have a private double and then a getter? The following is better:
public double Resolution { get; private set; }
It will create the field for you!

File: src/DroneSafetyApi/Models/Heatmap.cs [hunk 12]
```diff
@@ -1,12 +1,81 @@
...
 using System.Collections.Generic;
 using System.Linq;
+using System.Text;
 using System.Threading.Tasks;
 
 namespace DroneSafetyApi.Models
 {
-    public class Heatmap
+
+    public class HeatMap
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Models/Heatmap.cs:11]: I think that the concept of a "Model" has been lost. A model is something used to store the state program, in a way that is easy to display or send over an API. From this perspective, the Heatmap model should contain just the information you need to transmit over the API (keep it short). This class should certainly NOT have any custom JSON serialisation logic.

I think the Heatmap generation has been mixed up with the Heatmap model. This should be two separate classes: one in Models to represent the data, and one in Services to populate this model.

File: src/DroneSafetyApi/Models/Heatmap.cs [hunk 13]
```diff
@@ -1,12 +1,81 @@
...
+                    list.Add(jsonObject);
+                }
+            return JsonConvert.SerializeObject(list);
+        }
+
+
+
     }
-}
+}
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Models/Heatmap.cs:106]: Don't delete newlines at the end of the file. It's good practice to keep them (for backwards compatibility reasons apparently), and it also messes up the git history with unnecessary changes.

File: src/DroneSafetyApi/Models/IDataPointRepository.cs [hunk 14]
```diff
@@ -6,10 +6,10 @@
 
 namespace DroneSafetyApi.Models
 {
-    public interface IHazardsRepository
+    public interface IDataPointRepository
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Models/IDataPointRepository.cs:9]: DataPoint should be in the plural. This change should be propagated to all relevant places :)

File: src/DroneSafetyApi/Models/IDataPointRepository.cs [hunk 15]
```diff
@@ -6,10 +6,10 @@
...
     {
         // Keys in the return value signify the names of each relevant data source
         // Values in the return value signify the hazards intersecting in the given area
-        Dictionary<string, IEnumerable<Hazard>> GetHazardsOverlappingWith(Polygon area);
+        IEnumerable<DataPoint> GetDataPoints(double x, double y);
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Models/IDataPointRepository.cs]: What you you mean by "GetDataPoints"? Would be nice to have a slightly more descriptive method name.

File: src/DroneSafetyApi/Services/DataPointsToHeatmapResponse.cs [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+using DroneSafetyApi.Models;
+using Microsoft.Azure.Documents.Spatial;
+
+namespace DroneSafetyApi.Services
+{
+    public class DataPointsToHeatmapResponse : DataPointsToHeatmapsResponseNoCompositionSendAllSources
+    {
+
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Services/DataPointsToHeatmapResponse.cs:12]: Why so many blank lines?

File: src/DroneSafetyApi/Models/Heatmap.cs [hunk 17]
```diff
@@ -1,12 +1,81 @@
...
+        private double mStartX;
+        private double mStartY;
+        private int mDecimalPlaces;
+        private double mResolution;
+
+        public double GetResolution()
+        {
+            return mResolution;
+        }
+        public HeatMap(double minX, double maxX, double minY, double maxY, int decimalPlaces)
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Models/Heatmap.cs:25]: Why is specifying decimalPlaces the best way of defining the resolution?

File: src/DroneSafetyApi/Services/DataPointsToHeatmapResponse.cs [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+            // Calculate scaling between grid and lat long coordinates
+            double xScale = (area.Max.Latitude - area.Min.Latitude) / width;
+            double yScale = (area.Max.Longitude - area.Min.Longitude) / height;
+
+            foreach (DataPoint hazard in hazards)
+            {
+                IList<Position> coord = hazard.Area.Rings[0].Positions;
+
+                // Calculate bounding box of each hazard polygon
+                double minLong = coord[0].Longitude;
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Services/DataPointsToHeatmapResponse.cs]: Doesn't minLong = maxLong in the current code? Or is this like a pop operation?

File: src/DroneSafetyApi/Services/DataPointsToHeatmapResponse.cs [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+
+
+        public override HeatMap ConvertToHeatmap(int decimalPlaces, BoundingBox area, IEnumerable<DataPoint> hazards)
+        {
+            // Data Points 
+            // Get Type
+            // 
+
+            // Initialise Heatmap
+            HeatMap heatmap = new HeatMap(area.Min.Latitude, area.Max.Latitude, area.Min.Longitude, area.Max.Longitude, decimalPlaces);
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Services/DataPointsToHeatmapResponse.cs]: Make this multiple lines

File: src/DroneSafetyApi/Services/DataPointsToHeatmapResponse.cs [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+
+
+           
+
+            return heatmap;
+        }
+
+
+
+        private void ProcessPolygon(Polygon polygon, HeatMap heatmap)
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Services/DataPointsToHeatmapResponse.cs]: This method is 40 lines long! Replace with calls to submethods.

Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Services/DataPointsToHeatmapResponse.cs]: I trust you've tested the algorithm and it works nicely :p

File: src/DroneSafetyApi/Services/DataPointsToHeatmapResponse.cs [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+                    {
+                        heatmap.Values[xStart + x, height - 1 - (yStart + y)] += (inHazard(minLat + (x * xScale),
+                            minLong + (y * yScale), hazard)) ? hazard.Severity : 0;
+                    }
+                }
+            }
+        }
+
+
+        public Boolean inHazard(double lat, double lon, DataPoint hazard)
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Services/DataPointsToHeatmapResponse.cs]: Why are you returning a Boolean as opposed to a 'bool'?

File: src/DroneSafetyApi/Services/DataPointsToHeatmapsResponseNoCompositionSendAllSources.cs [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+namespace DroneSafetyApi.Services
+{
+    public abstract class DataPointsToHeatmapsResponseNoCompositionSendAllSources : IHazardsToHeatmapsResponse
+    {
+        public HeatmapsResponse ConvertToHeatmapResponse(
+            int decimalPlaces,
+            BoundingBox area,
+            IEnumerable<DataPoint> datapoints)
+        {
+            Dictionary<string, List<DataPoint>> datapointdictionary = new Dictionary<string, List<DataPoint>>();
```
Comment (dhruvtapasvi) [inline src/DroneSafetyApi/Services/DataPointsToHeatmapsResponseNoCompositionSendAllSources.cs:17]: Make it an IEnumerable, not a list (if this occurs anywhere else there too)

File: src/DroneSafetyApi/Models/IDataPointRepository.cs (hunk 14 above)
Comment (joshuablake) [reply src/DroneSafetyApi/Models/IDataPointRepository.cs:9]: Disagree, more natural to use the singular. Eg a box for books is a "book box" not a "books box".

File: src/DroneSafetyApi/Services/DataPointsToHeatmapsResponseNoCompositionSendAllSources.cs (hunk 22 above)
Comment (ranzmigrod) [reply src/DroneSafetyApi/Services/DataPointsToHeatmapsResponseNoCompositionSendAllSources.cs:17]: It needs to be a list in this case, otherwise we get into all kinds of problems

File: src/DroneSafetyApi/Models/HeatMapPoint.cs (hunk 9 above)
Comment (ranzmigrod) [reply src/DroneSafetyApi/Models/HeatMapPoint.cs:12]: There had to be a complete restructure of things, it was necessary

File: src/DroneSafetyApi/Models/Heatmap.cs (hunk 10 above)
Comment (ranzmigrod) [reply src/DroneSafetyApi/Models/Heatmap.cs:13]: Why not?

File: src/DroneSafetyApi/Models/Heatmap.cs (hunk 12 above)
Comment (ranzmigrod) [reply src/DroneSafetyApi/Models/Heatmap.cs:11]: I agree only that the generation of heatmap points and json converter should be somewhere else, the others are part of the model and would lead to annoying code otherwise

File: src/DroneSafetyApi/Models/Heatmap.cs (hunk 17 above)
Comment (Mudit-G) [reply src/DroneSafetyApi/Models/Heatmap.cs:25]: It isn't but it is ok for now until we get something actually working.

File: src/DroneSafetyApi/Models/Heatmap.cs (hunk 13 above)
Comment (Mudit-G) [reply src/DroneSafetyApi/Models/Heatmap.cs:106]: I think its a visual studio formatting thing.

File: src/DroneSafetyApi/Models/IDataPointRepository.cs (hunk 14 above)
Comment (Mudit-G) [reply src/DroneSafetyApi/Models/IDataPointRepository.cs:9]: Agree with Josh.

File: src/DroneSafetyApi/Models/HeatMapPoint.cs (hunk 8 above)
Comment (Mudit-G) [reply src/DroneSafetyApi/Models/HeatMapPoint.cs:10]: I think the Caesium heat map uses x and y but will confirm with Bence. If so it just avoids an extra processing step on the front end.

File: src/DroneSafetyApi/Models/HeatMapPoint.cs (hunk 9 above)
Comment (dhruvtapasvi) [reply src/DroneSafetyApi/Models/HeatMapPoint.cs:12]: I meant blank lines

File: src/DroneSafetyApi/Models/Heatmap.cs (hunk 10 above)
Comment (dhruvtapasvi) [reply src/DroneSafetyApi/Models/Heatmap.cs:13]: It's not happening anywhere else. Also, this notation is redundant with things like properties, it doesn't help

File: src/DroneSafetyApi/Models/Heatmap.cs (hunk 13 above)
Comment (dhruvtapasvi) [reply src/DroneSafetyApi/Models/Heatmap.cs:106]: Visual studio created the files with extra blank line so I'm sure this setting can be fixed

File: src/DroneSafetyApi/Services/DataPointsToHeatmapsResponseNoCompositionSendAllSources.cs (hunk 22 above)
Comment (dhruvtapasvi) [reply src/DroneSafetyApi/Services/DataPointsToHeatmapsResponseNoCompositionSendAllSources.cs:17]: Like?

Comment (dhruvtapasvi) [reply src/DroneSafetyApi/Services/DataPointsToHeatmapsResponseNoCompositionSendAllSources.cs:17]: Like?

File: src/DroneSafetyApi/Models/Heatmap.cs (hunk 10 above)
Comment (Mudit-G) [reply src/DroneSafetyApi/Models/Heatmap.cs:13]: We should agree on some conventions at some point but for now we should get something working.
